### PR TITLE
Add rest and transport plumbing

### PR DESCRIPTION
**Pull request by hub-cap**
Also wired up a bit more of the plugin and tested against a live
cluster.

### CODE REVIEW

**Comment by hub-cap** [comment]
This is a work in progress. its not close to done, but I wanted to get something that other people can work against. I still need some tests at the very minimum. I also want to make sure the naming of things is good, so thats something I would like to discuss as well. I am also not 100% sure what label to put it under.

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/rest/RestGenEnrichPolicyAction.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.rest.BaseRestHandler;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.action.RestToXContentListener;
+import org.elasticsearch.xpack.core.enrich.GetEnrichPolicyAction;
+
+import java.io.IOException;
+
+public class RestGenEnrichPolicyAction extends BaseRestHandler {
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/rest/RestGenEnrichPolicyAction.java]: RestGenEnrichPolicyAction -> RestGetEnrichPolicyAction

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/rest/RestGenEnrichPolicyAction.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    public RestGenEnrichPolicyAction(final Settings settings, final RestController controller) {
+        super(settings);
+        controller.registerHandler(RestRequest.Method.GET, "/{index}/_enrich", this);
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/rest/RestGenEnrichPolicyAction.java]: I think we should have the following path structure: `/_enrich/policy/{source_index}`

**Comment by martijnvg** [comment]
Should we adding this api to the enrich plugin instead? I think the only reason that we would want to add this to xpack-core is for transport client support (at least we did in the past iirc)? I don't think we should be adding transport client support and just hlrc support, so we can keep this inside the enrich plugin instead?

**Comment by droberts195** [comment]
> I think the only reason that we would want to add this to xpack-core is for transport client support

What would happen if another X-Pack plugin wanted to use this functionality?  For example, at the moment ML calls the `has_privileges` API in security, and only needs a compile time dependency on the `x-pack-core` plugin, not `x-pack-security`.  So if in the future ML wanted to call an enrich action and the actions were in the `x-pack-enrich` plugin it would need to depend on `x-pack-enrich`.  But then years later suppose that `x-pack-ml` and `x-pack-enrich` wanted to depend on different versions of some other 3rd party library.  Is this starting to erode the benefits of splitting up X-Pack?

This is all very theoretical at the moment, but may be something to consider.

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/rest/RestGenEnrichPolicyAction.java (hunk 2 above)
Comment (hub-cap) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/rest/RestGenEnrichPolicyAction.java]: ok, so completely remove the `{index}` and only put it in the payload of the policy itself, right?

**Comment by hub-cap** [comment]
> Should we adding this api to the enrich plugin instead? I think the only reason that we would want to add this to xpack-core is for transport client support (at least we did in the past iirc)? I don't think we should be adding transport client support and just hlrc support, so we can keep this inside the enrich plugin instead?

Since transport client, even tho its deprecated, is supported for all of 7, wouldn't that mean that we should support all of the calls that we can do even if the HLRC ones exist? I recently told @dimitris-athanasiou this was the case with some new stuff he was working on, so im CC'ing him on this :)

**Comment by martijnvg** [comment]
> Since transport client, even tho its deprecated, is supported for all of 7, wouldn't that mean that we should support all of the calls that we can do even if the HLRC ones exist?

I don't think we have to support new APIs in the transport client, now that it has been deprecated and because there is an alternative.

**Comment by hub-cap** [comment]
> there is an alternative.

Yes but the alternative might require a complete app rewrite if you want to use the rest client instead of the transport client for the entire 7.x series, or have 2 different ways to access a cluster depending on the APIs that you care to use. This seems painful for our users.

**Comment by jakelandis** [comment]
> This seems painful for our users.

The alternative is allow user to add new code against an API that will taken away in 8.0.0. There will be pain either way. a) code against deprecated now and re-write for 8.0.0 or b) start using the HLRC now.  I think the former is actually more painful. 

This argument is only for net-new features added in 7.x (like this one). I do believe that minor changes for features pre-existing in 6.x should continue support for the TransportClient.

**Comment by hub-cap** [comment]
superseded by #41383, #41384

**Pull request closed without merging**